datastore/mqtt/mqtt.go

```diff
@@ -0,0 +1,208 @@
+		airbrakeHandler:    airbrakeHandler,
+		namespace:          namespace,
+		ctx:                ctx,
+		ackChan:            ackChan,
+		reliableAckTxTypes: reliableAckTxTypes,
+	}, nil
+}
+
+func (p *MQTTProducer) Produce(rec *telemetry.Record) {
+	if p.ctx.Err() != nil {
+		return
+	}
+
+	// Alerts and errors are not published to MQTT
+	if rec.TxType == "alerts" || rec.TxType == "errors" {
+		p.logger.Log(logrus.DEBUG, "mqtt_ignored_record", logrus.LogInfo{"tx_type": rec.TxType})
+		return
+	}
+
+	payload, err := rec.GetProtoMessage()
+	if err != nil {
+		p.ReportError("mqtt_payload_unmarshal_error", err, nil)
+		return
+	}
+
+	data, ok := payload.(*protos.Payload)
+	if !ok {
+		p.ReportError("mqtt_invalid_payload_type", fmt.Errorf("expected *protos.Payload, got %T", payload), nil)
+		return
+	}
+
+	convertedPayload := transformers.PayloadToMap(data, false, p.logger)
+
+	var tokens []pahomqtt.Token
+
+	// Publish each field to the relevant topic on the MQTT broker
+	for key, value := range convertedPayload {
+		// Skip VIN and CreatedAt fields
+		if key == "Vin" || key == "CreatedAt" {
+			continue
+		}
+
+		topicName := fmt.Sprintf("%s/vin/%s/%s", p.config.TopicBase, rec.Vin, key)
+
+		jsonValue, err := json.Marshal(map[string]interface{}{"value": value})
+		if err != nil {
+			p.ReportError("mqtt_json_marshal_error", err, logrus.LogInfo{"key": key})
+			continue
+		}
+
+		token := p.client.Publish(topicName, p.config.QoS, p.config.Retained, jsonValue)
+		tokens = append(tokens, token)
+
+		metricsRegistry.byteTotal.Add(int64(len(jsonValue)), map[string]string{"record_type": rec.TxType})
+		metricsRegistry.publishCount.Inc(map[string]string{"record_type": rec.TxType})
+	}
+
+	var publishError bool
+	// Wait for all topics to be published
+	for _, token := range tokens {
+		if !token.WaitTimeout(time.Duration(p.config.PublishTimeout) * time.Second) {
+			metricsRegistry.errorCount.Inc(map[string]string{"record_type": rec.TxType})
+			p.ReportError("mqtt_publish_timeout", fmt.Errorf("publish operation timed out"), logrus.LogInfo{"topic": "multiple"})
+			publishError = true
+		} else if token.Error() != nil {
+			metricsRegistry.errorCount.Inc(map[string]string{"record_type": rec.TxType})
+			p.ReportError("mqtt_dispatch_error", token.Error(), logrus.LogInfo{"topic": "multiple"})
+			publishError = true
+		}
+	}
+
+	// Only process reliable ACK if no errors were reported
+	if !publishError {
+		p.ProcessReliableAck(rec)
+	}
+}
+
+func (p *MQTTProducer) ProcessReliableAck(entry *telemetry.Record) {
+	_, ok := p.reliableAckTxTypes[entry.TxType]
+	if ok {
+		p.ackChan <- entry
+		metricsRegistry.reliableAckCount.Inc(map[string]string{"record_type": entry.TxType})
+	}
+}
+
+func (p *MQTTProducer) ReportError(message string, err error, logInfo logrus.LogInfo) {
+	p.airbrakeHandler.ReportLogMessage(logrus.ERROR, message, err, logInfo)
+	p.logger.ErrorLog(message, err, logInfo)
+}
+
+func (p *MQTTProducer) Close() error {
+	p.client.Disconnect(250)
```

Comment:
nit: make `250` a configurable value?

Reply:
Yes, that is a good idea. And I'll set the default to 250ms.